internal/resource.go

```diff
@@ -1,140 +1,265 @@
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *Resource) downloadFromCache(dir string, mode os.FileMode, ctx context.Context) error {
+	token := os.Getenv("GRABIT_ARTIFACTORY_TOKEN")
+	if token == "" {
+		return fmt.Errorf("GRABIT_ARTIFACTORY_TOKEN must be set for cache operations")
+	}
+
+	log.Debug().Str("cache", r.CacheUri).Msg("Attempting cache download")
+	tempFile, err := downloadWithToken(r.CacheUri, dir, ctx, token)
+	if err != nil {
+		return err
+	}
+
+	if err := checkIntegrityFromFile(tempFile, "sha256", r.Integrity, r.CacheUri); err != nil {
+		os.Remove(tempFile)
+		log.Debug().Msg("Cache integrity validation failed")
 		return err
 	}
-	var downloadError error = nil
-	for _, u := range l.Urls {
-		// Download file in the target directory so that the call to
-		// os.Rename is atomic.
-		lpath, err := GetUrlToDir(u, dir, ctx)
+
+	return renameAndSetPermission(tempFile, filepath.Join(dir, getLocalFileName(r.CacheUri)), mode)
+}
+
+func (r *Resource) downloadFromSource(dir string, mode os.FileMode, ctx context.Context) error {
+	var lastErr error
+	for _, u := range r.Urls {
+		log.Debug().Str("url", u).Msg("Attempting download from source")
+		tempFile, err := GetUrlToDir(u, dir, ctx)
 		if err != nil {
-			downloadError = err
+			lastErr = err
+			log.Debug().Err(err).Msg("Download failed, trying next URL")
 			continue
 		}
-		err = checkIntegrityFromFile(lpath, algo, l.Integrity, u)
-		if err != nil {
+
+		if err := checkIntegrityFromFile(tempFile, "sha256", r.Integrity, u); err != nil {
 			return err
 		}
 
-		localName := ""
-		if l.Filename != "" {
-			localName = l.Filename
-		} else {
+		localName := r.Filename
+		if localName == "" {
 			localName = path.Base(u)
 		}
-		resPath := filepath.Join(dir, localName)
-		err = os.Rename(lpath, resPath)
-		if err != nil {
-			return err
-		}
-		if mode != NoFileMode {
-			err = os.Chmod(resPath, mode.Perm())
-			if err != nil {
-				return err
-			}
-		}
-		ok = true
+
+		return renameAndSetPermission(tempFile, filepath.Join(dir, localName), mode)
 	}
-	if !ok {
-		if downloadError != nil {
-			return downloadError
-		}
+
+	return fmt.Errorf("failed to download from any source: %v", lastErr)
+}
+
+func (r *Resource) uploadToCache(dir string, ctx context.Context) error {
+	token := os.Getenv("GRABIT_ARTIFACTORY_TOKEN")
+	if token == "" {
+		return fmt.Errorf("GRABIT_ARTIFACTORY_TOKEN must be set for cache operations")
+	}
+
+	filePath := filepath.Join(dir, getLocalFileName(r.CacheUri))
+	hash, err := GetFileHash(filePath)
+	if err != nil {
+		return err
+	}
+
+	// Clean the base URI and use hash as filename
+	baseUri := strings.TrimSuffix(r.CacheUri, "/")
+	cachePath := fmt.Sprintf("%s/%s", baseUri, hash)
+
+	log.Debug().
+		Str("cachePath", cachePath).
+		Msg("Uploading to cache")
+
+	fileData, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return err
+	}
+
+	// Use simple http.NewRequest for more control over the request
```

Comment:
What does this comment mean?  "more control" compared to what?

Reply:
Using http.NewRequest instead of higher-level HTTP functions gives more detailed control over the request sent to the Artifactory server. This allows the code to set headers, the request body, and other details directly, which might be needed for Artifactory. This level of control is important to make sure the request is formatted correctly for Artifactory. While higher-level libraries are simpler to use, http.NewRequest gives the flexibility needed to work with Artifactory in this case.

Reply:
The higher-level libraries are not just simpler to use, they are much safer and that's why we should use them. See https://github.com/earthboundkid/requests?tab=readme-ov-file#faqs for details.

Can you name *one thing* that the http.NewRequest can do that github.com/carlmjohnson/requests cannot do? If not, let's use github.com/carlmjohnson/requests.

For setting headers, see https://github.com/earthboundkid/requests/wiki#custom-headers